src/cargo/ops/cargo_add/mod.rs

```diff
@@ -124,11 +126,71 @@ pub fn add(workspace: &Workspace<'_>, options: &AddOptions<'_>) -> CargoResult<(
                 inherited_features.iter().map(|s| s.as_str()).collect();
             unknown_features.extend(inherited_features.difference(&available_features).copied());
         }
+
         unknown_features.sort();
+
+        let mut activated: IndexSet<_> =
+            dep.features.iter().flatten().map(|s| s.as_str()).collect();
+        if dep.default_features().unwrap_or(true) {
+            activated.insert("default");
+        }
         if !unknown_features.is_empty() {
-            anyhow::bail!("unrecognized features: {unknown_features:?}");
+            let (activated, deactivated) = dep.features();
+            let deactivated = deactivated
+                .iter()
+                .filter(|f| !unknown_features.contains(f))
+                .collect::<Vec<_>>();
+            let activated = activated
+                .iter()
+                .filter(|f| !unknown_features.contains(f))
+                .collect::<Vec<_>>();
+            let mut message = format!(
+                "unrecognized feature{} for crate {}: {}\n",
+                if unknown_features.len() == 1 { "" } else { "s" },
+                dep.name,
+                unknown_features.iter().format(", "),
+            );
+            if activated.is_empty() && deactivated.is_empty() {
+                write!(message, "no features available for crate {}", dep.name)?;
+            } else {
+                if !deactivated.is_empty() {
+                    writeln!(
+                        message,
+                        "disabled features:\n    {}",
+                        deactivated
+                            .iter()
+                            .map(|s| s.to_string())
+                            .coalesce(|x, y| if x.len() + y.len() < 78 {
+                                Ok(format!("{x}, {y}"))
+                            } else {
+                                Err((x, y))
+                            })
+                            .into_iter()
+                            .format("\n    ")
+                    )?
+                }
+                if !activated.is_empty() {
+                    writeln!(
+                        message,
+                        "enabled features:\n    {}",
+                        activated
+                            .iter()
+                            .map(|s| s.to_string())
+                            .coalesce(|x, y| if x.len() + y.len() < 78 {
+                                Ok(format!("{x}, {y}"))
+                            } else {
+                                Err((x, y))
+                            })
+                            .into_iter()
+                            .format("\n    ")
+                    )?
+                }
+            }
+            anyhow::bail!(message.trim().to_owned());
         }
 
+        print_dep_table_msg(&mut options.config.shell(), &dep)?;
```

Comment:
nit: ideally the splitting of the prints would be at least a separate commit from the new error message as these are related but separate changes.  If nothing else, this makes it easier on the reviewer.

Reply:
good point, if you want I can still split it out

Reply:
I've already gone through the review of it here so not a blocker (would be if this were bigger) but up to you and for consideration for future changes.